### PR TITLE
Generalize artifacts images, fixes machine-os-content build flow and minor fixes

### PR DESCRIPTION
**Pull request by aleskandro**
Hi @vrutkovs @LorbusChris , this PR adds a new build-config to make an image similar to the origin's `artifacts` that can be used as we do in the coreos tekton pipeline (see https://github.com/okd-project/okd-coreos-pipeline/pull/2/files) to pull the RPMs for oc and hyperkube.

Since there already is a "hyperkube-rpms" image, we could also move the rpm build of oc there and change its name from `hyperkube-rpms` to `artifacts`. What do you think?

### CODE REVIEW

File: buildconfigs/20-artifacts.yaml
Comment (LorbusChris) [inline buildconfigs/20-artifacts.yaml:5]:
```suggestion
  namespace: okd
```

File: forks/artifacts/Dockerfile.centos9
Comment (LorbusChris) [inline forks/artifacts/Dockerfile.centos9:2]: why is this not 4.12 here?

File: buildconfigs/20-artifacts.yaml
Comment (aleskandro) [reply buildconfigs/20-artifacts.yaml:5]: 🤦‍♀️

File: forks/artifacts/Dockerfile.centos9
Comment (aleskandro) [reply forks/artifacts/Dockerfile.centos9]: I set the value to be the same as our default one in Prow. Changed now.

File: forks/artifacts/Dockerfile.centos9
Comment (LorbusChris) [inline forks/artifacts/Dockerfile.centos9:16]: The way this is done in Prow CI is just awful, let's do it properly here:
```suggestion
RUN yum install -y --setopt=skip_missing_names_on_install=False \
        rpm-build \
        createrepo \
        bsdtar \
        krb5-devel \
        rsync \
    && TMPDIR=$( mktemp -d ) \
    && RPMBUILD_DIR=_output/local/releases \
    && mkdir -p ${RPMBUILD_DIR}/{BUILD,SOURCES,SPECS,SRPMS} \
    && SOURCE_GIT_COMMIT=$(git rev-parse --short "HEAD^{commit}" 2>/dev/null) \
    && SOURCE_GIT_TAG=$(git describe --long --tags --abbrev=7 --match '*[0-9]*' | sed -E 's/openshift-clients-([0-9]+\.[0-9]+\.[0-9]+)-.*/\1/' || echo '0.0.0-unknown-${SOURCE_GIT_COMMIT}') \
    && SOURCE_GIT_TREE_STATE=$(( ( [ ! -d ".git/" ] || git diff --quiet ) && echo 'clean' ) || echo 'dirty') \
   #&& SOURCE_GIT_MAJOR= \
   #&& SOURCE_GIT_MINOR= \
  && RPM_VERSION=$(set -o pipefail && echo '${SOURCE_GIT_TAG}' | sed -E 's/v([0-9]+\.[0-9]+\.[0-9]+)-.*/\1/').$(shell date -u +'%Y%m%d%H%M').${SOURCE_GIT_COMMIT} \
  && tar -czf "${TMPDIR}/openshift-clients-${RPM_VERSION}.tar.gz" \
    --exclude='.git' \
    --transform='s|^.|openshift-clients-${RPM_VERSION}|' \
    --exclude='openshift-clients-${RPM_VERSION}.tar.gz' \
    ./ \
  && cp -ivh ./oc.spec "${RPMBUILD_DIR}/SPECS/" \
  && cp -ivh "${TMPDIR}/openshift-clients-${RPM_VERSION}.tar.gz" "${RPMBUILD_DIR}/SOURCES/" \
  # TODO: add keys and --sign
  && rpmbuild -ba --nodeps \
    --define "_topdir $(pwd)/${RPMBUILD_DIR}" \
    --define "_rpmdir $(pwd)/${RPMBUILD_DIR}/rpms" \
    --define 'version ${RPM_VERSION}' \
    --define "os_git_vars OS_GIT_VERSION='${SOURCE_GIT_TAG}' OS_GIT_COMMIT='${SOURCE_GIT_COMMIT}' OS_GIT_MAJOR='${SOURCE_GIT_MAJOR}' OS_GIT_MINOR='${SOURCE_GIT_MIJOR}' OS_GIT_TREE_STATE='${SOURCE_GIT_TREE_STATE}'" \
    ${RPMBUILD_DIR}/SPECS/oc.spec
```

File: forks/artifacts/Dockerfile.centos9
Comment (aleskandro) [reply forks/artifacts/Dockerfile.centos9:20]: I don't think so (otherwise, similar issues we discussed in the past days would have been assessed)... we could use TARGETARCH or BUILDARCH though, let me check and I'll change it

File: forks/artifacts/Dockerfile.centos9
Comment (aleskandro) [reply forks/artifacts/Dockerfile.centos9:16]: :+1:

File: forks/artifacts/Dockerfile.centos9
Comment (aleskandro) [reply forks/artifacts/Dockerfile.centos9]: /hold

**Comment by vrutkovs** [comment]
Not sure why `okd-rpms` would not work. Do we need it for multiarch?

`artifacts` is too generic, lets add smth with `rpms` there

File: buildconfigs/01-imagestream.yaml
Comment (LorbusChris) [inline buildconfigs/01-imagestream.yaml:10]: let's get of this tag here, it's not public anymore

**Comment by LorbusChris** [comment]
Using `okd-` anything is not helping with alignment with ocp. I think artifacts is fine imo, and we just have to use it properly. E.g. everything (tars, bins and rpms) should be in there, (and in Prow, as long as we don't manifestlist support, probably ship multi-arch bins too).
So we way I see this it should be:
```
hyperkube-artifacts 
| cli-artifacts
| |
V V
artifacts
```

**Comment by vrutkovs** [comment]
> So we way I see this it should be:

Right, that looks good to me.

File: forks/artifacts/Dockerfile.centos9
Comment (aleskandro) [reply forks/artifacts/Dockerfile.centos9:39]: Reintroduced the RUN layer as discussed...

**Comment by aleskandro** [comment]
Hello, I'm trying to generalize the `*-artifacts` images to have them buildable from a single "dockerfile".

Based on what we disccused, it would be nice to have a single dockerfile to build multiarch binaries and RPMs for both oc and hyperkube.

Doing it for the RPMs is ok. However, binaries are built differently:

1. hypershift: KUBE_BUILD_PLATFORMS="...." make WHAT...
2. oc: make (cross-)build 

The output in the build layer would be different, but manageable:
1. hypershift: _output/local/bin/linux/%arch%/{bins}
2. oc: _output/local/bin/%os%_%arch%

The runtime stage will also need some trick to have the final multi-arch binaries in the place where their consumers look for them. See the [cli-artifacts dockerfile](https://github.com/openshift/oc/blob/af05172ab6dbb11dd3309268cf486fb3e482e7ce/images/cli-artifacts/Dockerfile.rhel) to have an idea.

**Solution1**: Generlize by using an arg that provides the whole command to run the build. 
In this case, we'd build twice for each architecture, rpm + single binary. It seems not elegant and slow to me. Isn't it?

**Solution2**: We'd have to extract each rpm for each architecture, get the executables and place them accordingly in their desired final dir for the payload's image.
The build will be faster as there is no double build, and no "eval-like" code has to be provided in a build argument. 

The drawback is for `oc`. When the arch is x86_64, the rpmbuild command creates two RPMs, one having the oc binary for win and x86_64 mac.
Then, we'd loose the darwin/arm64 binary. <- something should change here eventually, but we cannot make this solution the upstream one currently.

Solution3: Put the script to generalize the RPM building process in an auxiliary image (the builder imo) and continue maintaining two different and smaller Dockerfiles, hyperkube-artifacts and cli-artifacts, to join together in the artifacts one.

We still use the double build process, but since the Dockerfiles need to be kept in two repos, we will have a more homogeneous building process. The double build could be slightly improved in performance maybe.

Do you have any hint?

File: forks/cli-artifacts/Dockerfile.centos9
Comment (aleskandro) [inline forks/cli-artifacts/Dockerfile.centos9:21]: Hello, I'm trying to generalize the `*-artifacts` images to have them buildable from a single "dockerfile".

Based on what we disccused, it would be nice to have a single dockerfile to build multiarch binaries and RPMs for both oc and hyperkube.

Doing it for the RPMs is ok. However, binaries are built differently:

1. hypershift: KUBE_BUILD_PLATFORMS="...." make WHAT...
2. oc: make (cross-)build 

The output in the build layer would be different, but manageable:
1. hypershift: _output/local/bin/linux/%arch%/{bins}
2. oc: _output/local/bin/%os%_%arch%

The runtime stage will also need some trick to have the final multi-arch binaries in the place where their consumers look for them. See the [cli-artifacts dockerfile](https://github.com/openshift/oc/blob/af05172ab6dbb11dd3309268cf486fb3e482e7ce/images/cli-artifacts/Dockerfile.rhel) to have an idea.

**Solution1**: Generlize by using an arg that provides the whole command to run the build. 
In this case, we'd build twice for each architecture, rpm + single binary. It seems not elegant and slow to me. Isn't it?

**Solution2**: We'd have to extract each rpm for each architecture, get the executables and place them accordingly in their desired final dir for the payload's image.
The build will be faster as there is no double build, and no "eval-like" code has to be provided in a build argument. 

The drawback is for `oc`. When the arch is x86_64, the rpmbuild command creates two RPMs, one having the oc binary for win and x86_64 mac.
Then, we'd loose the darwin/arm64 binary. <- something should change here eventually, but we cannot make this solution the upstream one currently.

Solution3: Put the script to generalize the RPM building process in an auxiliary image (the builder imo) and continue maintaining two different and smaller Dockerfiles, hyperkube-artifacts and cli-artifacts, to join together in the artifacts one.

We still use the double build process, but since the Dockerfiles need to be kept in two repos, we will have a more homogeneous building process. The double build could be slightly improved in performance maybe.

Do you have any hint?

Comment (LorbusChris) [reply forks/cli-artifacts/Dockerfile.centos9:21]: Great overview, thanks for providing that context! Sounds to me like Solution3 would be the sanest then

**Comment by aleskandro** [comment]
To clarify the output (I forgot to take the artifacts one, but it's trivial to derive). We are using a bit more space due to the (temporary) redundancy needed for "backward compatibility" with the consumers of these images.

# cli-artifacts

```
[root@cli-1 openshift]# tree /usr/share/openshift
.
├── LICENSE
├── darwom_amd64
│   └── oc
├── darwin_arm64
│   └── oc
├── linux_amd64
│   └── oc
├── linux_arm64
│   └── oc
├── linux_ppc64le
│   └── oc
├── linux_s390x
│   └── oc
├── mac
│   └── oc
├── mac_arm64
│   └── oc
├── windows
│   └── oc.exe
└── windows_amd64
    └── oc.exe

7 directories, 8 files

[root@cli-1 repo]# tree /srv/repo
/srv/repo
├── aarch64
│   └── openshift-clients-4.12.0.202209161546.5b29f86cf-1.el9.aarch64.rpm
├── ppc64le
│   └── openshift-clients-4.12.0.202209161546.5b29f86cf-1.el9.ppc64le.rpm
├── s390x
│   └── openshift-clients-4.12.0.202209161546.5b29f86cf-1.el9.s390x.rpm
└── x86_64
    ├── openshift-clients-4.12.0.202209161546.5b29f86cf-1.el9.x86_64.rpm
    └── openshift-clients-redistributable-4.12.0.202209161546.5b29f86cf-1.el9.x86_64.rpm

4 directories, 5 files
[root@cli-1 repo]# 


```

## Hyperkube

```

[root@built-hyperkube openshift]# tree /usr/share/openshift
/usr/share/openshift
├── LICENSE
└── linux
    └── arm64
        ├── conversion-gen
        ├── deepcopy-gen
        ├── defaulter-gen
        ├── go2make
        ├── kube-apiserver
        ├── kube-controller-manager
        ├── kube-scheduler
        ├── kubelet
        ├── openapi-gen
        ├── prerelease-lifecycle-gen
        └── watch-termination

2 directories, 12 files
[root@built-hyperkube openshift]# 

```

## cli

```

[root@built-cli-2 srv]# tree /srv
/srv
├── repo
│   └── aarch64
│       └── openshift-clients-4.12.0.202209161655.5b29f86cf-1.el9.aarch64.rpm
└── zips
    └── openshift-clients-4.12.0.202209161655.5b29f86cf.tar.gz

3 directories, 2 files

[root@built-cli-2 srv]# tree /usr/share/openshift/
/usr/share/openshift/
├── LICENSE
└── linux
    ├── clicheck
    ├── gendocs
    ├── genman
    └── oc

1 directory, 5 files
[root@built-cli-2 srv]# 

```

File: buildconfigs/30-machine-os-content.yaml
Comment (LorbusChris) [inline buildconfigs/30-machine-os-content.yaml:18]: Shouldn't this be `registry.ci.openshift.org/origin/4.12:artifacts`?

Comment (aleskandro) [reply buildconfigs/30-machine-os-content.yaml:18]: yes, let me verify even the rest. Thanks

**Comment by LorbusChris** [comment]
LGTM pending my comment above.
@aleskandro thanks for working on this! Feel free to hit the merge button on this after addressing

**Review by LorbusChris — APPROVED** [review]

**Comment by aleskandro** [comment]
Thanks Christian

**Pull request merged**
